File: blockchain/blockchain.go
package blockchain

import (
	"crypto/sha256"
	"fmt"
	"sync"
)

type block struct {
	Data     string
	Hash     string
	PrevHash string
}

type blockchain struct {
	blocks []*block // block 구조체의 주소를 저장하기 위해 (for 메모리 효율)
}

var bc *blockchain

var once sync.Once

func (b *block) calculateHash() {
	hash := sha256.Sum256([]byte(b.Data + b.PrevHash))
	b.Hash = fmt.Sprintf("%x", hash)
}

func (bc *blockchain) getLastHash() string {
	totalBlocks := len(bc.blocks)
	if totalBlocks == 0 {
		return ""
	}
	return bc.blocks[totalBlocks-1].Hash
}

func (bc *blockchain) createBlock(data string) *block {
	newBlock := block{data, "", bc.getLastHash()}
	newBlock.calculateHash()
	return &newBlock
}

func (bc *blockchain) AddBlock(data string) {
	bc.blocks = append(bc.blocks, bc.createBlock(data))
}

func GetBlockchain() *blockchain {
	if bc == nil {
		once.Do(func() {
			bc = &blockchain{}
			bc.AddBlock("Genesis Block")
		})

	}
	return bc
}

func (bc *blockchain) AllBlocks() []*block {
	return GetBlockchain().blocks
}

File: blockchain.js
const crypto = require("crypto");

class Block {
  constructor(data, prevHash, height) {
    this.data = data;
    this.prevHash = prevHash;
    this.hash = this.calculateHash();
    this.height = height;
  }

  calculateHash() {
    const hashString = this.data + this.prevHash;
    return crypto.createHash("sha256").update(hashString).digest("hex");
  }
}

class Blockchain {
  constructor() {
    if (!Blockchain.instance) {
      this.blocks = [];
      this.addBlock("Genesis Block");
      Blockchain.instance = this;
    }

    return Blockchain.instance;
  }

  getLastHash() {
    if (this.blocks.length === 0) {
      return "";
    }
    return this.blocks[this.blocks.length - 1].hash;
  }

  createBlock(data) {
    const prevHash = this.getLastHash();
    const height = this.blocks.length + 1;
    return new Block(data, prevHash, height);
  }

  addBlock(data) {
    const newBlock = this.createBlock(data);
    this.blocks.push(newBlock);
  }

  allBlocks() {
    return this.blocks;
  }

  getBlock(height) {
    if (height <= 0 || height > this.blocks.length) {
      return null;
    }
    return this.blocks[height - 1];
  }
}

const instance = new Blockchain();
Object.freeze(instance); // Prevent modification

module.exports = instance;
